psimol/main.py

```diff
@@ -452,7 +502,114 @@ def from_cif(cls, file_path: str) -> Molecule:
         Args:
             file_path (str): Path to the .cif file
         """
-        pass #TODO
+        
+        with open(file_path, 'r') as file:
+            lines = file.readlines()
+
+        # the parser will advance to the 'atom_site' loop of the .cif file
+        # (in STAR files (.cif included), tables are called "loops"),
+        # assemble the loop's contents, then parse them
+
+        # used to test if a line is from the header of the 'atom_site' loop
+        def _is_atom_site_header(s: str) -> bool:
+            return bool(re.match(r'_atom_site[_.]', s, re.IGNORECASE))
+        # parser state variables
+        in_atom_site = in_atom_site_body = False
+
+        # will hold the columns of the 'atom_site' loop, along with their order
+        atom_site_header: dict[str:int] = {}
+        ncol = 0
+        # columns required to be present by the parser
+        required_columns = ['type_symbol', 'cartn_x', 'cartn_y', 'cartn_z']
+        # will hold the body of the 'atom_site' loop
+        atom_site_body: list[str] = ""
+
+        for line in lines:
+            line = line.lstrip()
+
+            # phase 1: advance parser to 'atom_site' loop
+            # (in STAR files (CIF included), tables are called "loops")
```

Comment:
Perhaps we can extract such table with multiline regex from the whole file text (`file.read()`), so possibly we could diminish the number of iterations / make the code more compact, but it's optional.